lib/rubocop/cop/rspec/message_spies.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that message expectations are set using spies.
+      #
+      # This cop can be configured in your configuration using the
+      # `EnforcedStyle` option and supports `--auto-gen-config`.
+      #
+      # @example `EnforcedStyle: have_received`
+      #
+      #   # bad
+      #   expect(foo).to receive(:bar)
+      #
+      #   # good
+      #   expect(foo).to have_received(:bar)
+      #
+      # @example `EnforcedStyle: receive`
+      #
+      #   # bad
+      #   expect(foo).to have_received(:bar)
+      #
+      #   # good
+      #   expect(foo).to receive(:bar)
+      #
+      class MessageSpies < Cop
+        include RuboCop::RSpec::SpecOnly, ConfigurableEnforcedStyle
+
+        MSG_RECEIVE = 'Prefer `receive` for setting message ' \
+          'expectations.'.freeze
+        MSG_HAVE_RECEIVED = 'Prefer `have_received` for setting message ' \
+          'expectations. You will need to set up `%s` as a spy, using either ' \
+          '`allow` or `spy`.'.freeze
+
+        SUPPORTED_STYLES = %w(have_received receive).freeze
+
+        def_node_matcher :message_expectation, %(
+          (send (send nil :expect (send nil $_)) :to ...)
+        )
+
+        def_node_search :receive_message, %(
+          $(send nil {:receive :have_received} ...)
+        )
+
+        def on_send(node)
+          message_expectation(node) do |receiver|
+            receive_message(node) do |match|
```

Comment:
I would probably name `match` something like `message_matcher`. I usually try to dig deep into rspec's source to find proper naming for the parts of their API that I'm dealing with. In this case [`have_received`](https://github.com/rspec/rspec-mocks/blob/c0bf45b5ceb179c51b0d38d7c91faf60a7a6048f/lib/rspec/mocks/example_methods.rb#L281-L283) and [`received`](https://github.com/rspec/rspec-mocks/blob/c0bf45b5ceb179c51b0d38d7c91faf60a7a6048f/lib/rspec/mocks/syntax.rb#L115) both are dealing with messages and setting matchers so `message_matcher` seems good enough